### PR TITLE
[https://nvbugs/6571220][fix] Correct unfused attention context workspace sizing

### PR DESCRIPTION
**Pull request by pranav-nvidia**
## Description

`AttentionOp::getWorkspaceSizeForContext` (the size query) and `AttentionOp::enqueueContext` (the carve) disagreed on four unfused-MHA context buffers, so the query could return fewer bytes than the carve consumes and the carved views ran past the end of the allocation. The shortfall grows with batch size, so it presented as a batch-size threshold rather than a universal failure.

| buffer | size query | enqueue carve |
|---|---|---|
| `attentionMask` | `max_num_tokens * kv_seq_length` | `batch_size * input_seq_length * kv_seq_length` |
| `qkvBuf` | `max_num_tokens * local_hidden_units_qo` | `batch_size * input_seq_length * local_hidden_units_qo` |
| `paddingOffset` | `sizeof(int) * max_num_tokens` | `sizeof(int) * batch_size * input_seq_length` |
| `encoderPaddingOffset` | `sizeof(int) * max_num_tokens` | `sizeof(int) * batch_size * cross_kv_length` |

With padding removal the real token count is smaller than `batch_size * max(context q length)`, so sizing by `max_num_tokens` underestimates. Separately, `cpp/tensorrt_llm/thop/attentionOp.cpp` passed a literal `0` for `cross_kv_length`; for cross-attention `kv_seq_length` *is* `cross_kv_length`, so `attentionMask`, `kBuf`, `vBuf`, `qkBuf` and `qkFloatBuf` were queried at zero bytes while the carve sized them from the real encoder KV length.

Every affected buffer is guarded by `mEnableContextFMHA ? 0 : <size>`, so fused FP16/BF16 paths are unaffected. The models that reach it are those falling back to unfused MHA: **T5/mT5 at any dtype** (relative position embedding) and **anything at FP32**. BART and Whisper at FP16/BF16 take the fused path.

Two measured exposures:

- **FP32 encoder-decoder with decoder CUDA graphs** — `CUBLAS_STATUS_EXECUTION_FAILED` during warmup capture, then an illegal memory access. This is what the temporary guard in `model_engine.py` worked around; the guard is removed here.
- **T5 BF16 with encoder CUDA graphs at batch >= 4** — `CUBLAS_STATUS_INTERNAL_ERROR`. FP32 is not required to hit this.

Latent since ~`v0.11.0`, because nothing exercised the unfused encoder-decoder cross-attention path above batch 2.

The fix passes the context sequences' maximum past-KV length into the size query as `cross_kv_length` and aligns the remaining packed-token buffers with the dense layout the carve uses, so the query upper-bounds the carve.

Note on scope: three of the four `common/attentionOp.cpp` changes are a packed-vs-dense fix that also applies to *self*-attention with padding removal, not only to cross-attention.

## Test Coverage

- `test_llm_api_pytorch_t5.py` — new FP32 mixed-encoder-length, batch-2, KV-manager-v1 CUDA graph case, added to `l0_h100.yml`. The mixed-batch cases now assert decoder CUDA graphs actually captured, so an engine that silently declines capture fails instead of passing on output checks alone.
- `test_llm_api_pytorch_whisper.py` — the FP32 feature-combination case now expects decoder graph capture rather than the engine declining graphs.

Verified on SM120 with a single-variable A/B (same binary otherwise), `t5-small` mixed-encoder-length suite: the FP32 case fails without the fix and passes with it; the BF16 greedy, BF16 beam-2 and KV-v2 cases pass in both arms.

## PR Checklist

- [x] Please check this after reviewing the above items as appropriate for this PR.

## Dev Engineer Review

- Updated unfused attention workspace sizing to match `enqueueContext` buffer carving.
- Passed the maximum cross-KV length through the workspace query.
- Preserved fused context FMHA behavior.
- Removed the temporary FP32 encoder-decoder CUDA graph guard.
- Kept workspace API declarations consistent.
- Local checks passed.
- GPU CI validation remains unresolved. The initial run lacked the required `ci: full pre-merge approved` label, and the later L0 pipeline still failed.

## QA Engineer Review

- Added FP32 T5 mixed-encoder-length coverage in `test_llm_api_pytorch_t5.py`.
- Added assertions for decoder CUDA graph capture and disabled encoder graph capture.
- Updated Whisper FP32 graph-capture expectations in `test_llm_api_pytorch_whisper.py`.
- Added the T5 test to `tests/integration/test_lists/test-db/l0_h100.yml`.
- The T5 test is covered by the H100 CI test list.
- Verdict: needs follow-up. A successful GPU CI run must validate the T5 FP32 decoder graph case and the updated Whisper expectation.

### CODE REVIEW

**Comment by BowenFu** [comment]
Verified the root cause independently — the size query and the `enqueueContext` carve really had diverged, and this makes them match term for term.

`enqueueContext` carves (`attentionOp.cpp:1502, 1524, 1591-1594`):

- `attention_mask` = `batch_size * input_seq_length * kv_seq_length`
- `qkv_buf_2` = `batch_size * input_seq_length * local_hidden_units_qo`
- `paddingOffset` = `batch_size * input_seq_length`
- `encoderPaddingOffset` = `batch_size * cross_kv_length`

while `getWorkspaceSizeForContext` (`:799, 822, 897-898`) sized all four off `max_num_tokens`. With padding removal `num_tokens < batch_size * input_seq_length`, so the query under-allocated exactly as the removed guard's comment described. After this PR all four terms are identical to the carve.

One thing I checked so nobody else has to: `encoder_padding_offset_size` now collapses to 0 whenever `cross_kv_length == 0`, i.e. on every non-cross-attention op, where it used to be `sizeof(int) * max_num_tokens`. That's safe — `decoder_params.encoderPaddingOffsets` is `isCrossAttention() ? workspaceViews.encoderPaddingOffset : nullptr` (`:1670`), so the slice is never dereferenced there. And everything in this hunk is gated on `!mEnableContextFMHA`, so the fused default path is untouched.

Not approving yet, and only for want of evidence: the run at `699e2f7` (PR_Github #64109) came back FAILURE with **Multi-GPU Label Required**, so the multi-GPU stages never ran. That leaves the two cases that actually matter unproven — the new `fp32-kv-v1-decoder-cuda-graph-on-greedy-batch2` T5 case, and the Whisper `fp32-kv-v1-graphs-requested-greedy` expectation flipped from "declines graphs" to "captures". Dropping the guard re-enables decoder CUDA graphs for fp32 enc-dec, so those two are the only thing standing between this and a repeat of the cublas `EXECUTION_FAILED`. Worth getting the `ci: full pre-merge approved` label and a green run on this exact head.

File: cpp/tensorrt_llm/common/attentionOp.cpp [hunk 1]
```diff
@@ -799,7 +799,10 @@ size_t AttentionOp::getWorkspaceSizeForContext(tensorrt_llm::DataType type, int3
 
     auto const batch_size = static_cast<size_t>(max_num_seq);
     auto const kv_seq_length = (isCrossAttention() ? cross_kv_length : input_seq_length);
-    size_t const attention_mask_size = mEnableContextFMHA ? 0 : size * max_num_tokens * kv_seq_length;
+    // The unfused-MHA buffers below must upper-bound the enqueueContext carve, which sizes them by
+    // batch_size * input_seq_length (not num_tokens): with padding removal the actual token count can be
+    // smaller than batch_size * max(context q length), so sizing by max_num_tokens underestimates.
+    size_t const attention_mask_size = mEnableContextFMHA ? 0 : size * batch_size * input_seq_length * kv_seq_length;
```
Comment (brnguyen2) [inline cpp/tensorrt_llm/common/attentionOp.cpp:805]: This is the right fix, but it patches one of two independently written copies of the same arithmetic — `enqueueContext` recomputes all ~25 of these sizes at line 1504 onward, and both then fill an `AttentionContextWorkspaceSizes` and call `buildContextLayout`. The bug being fixed here *is* that drift, and nothing prevents the next one: no assert compares the planned size against the carved layout.

Since both sides already funnel through the same struct, a follow-up that extracts one `computeContextWorkspaceSizes(batch, input_seq_length, cross_kv_length, num_tokens, total_kv_len, ...)` called from both — the size query passing engine maxima, the enqueue passing runtime dims — would make this class of bug structurally impossible. Failing that, a `TLLM_CHECK(carvedTotal <= plannedTotal)` in `enqueueContext` would at least turn a silent OOB into a clear error instead of `CUBLAS_STATUS_EXECUTION_FAILED`.

File: cpp/tensorrt_llm/thop/attentionOp.cpp [hunk 2]
```diff
@@ -1360,8 +1362,17 @@ void attention(torch::Tensor q, std::optional<torch::Tensor> k, std::optional<to
...
-    int64_t const workspace_size = runner->getWorkspaceSize(
-        *op, num_tokens, max_attention_window_size, num_gen_tokens, max_blocks_per_sequence, ctx_total_kv_len);
+    // For cross-attention, several unfused-path context buffers scale with the encoder KV length.
+    // Mirror the context-stage enqueue, which uses the max past-KV length over the context sequences
+    // as cross_kv_length; sizing with 0 here under-allocates the workspace and the carved views in
+    // enqueueContext land past the end of the allocation.
+    int32_t max_cross_kv_length = 0;
+    if (op->isCrossAttention() && num_contexts > 0)
+    {
+        max_cross_kv_length = host_past_key_value_lengths.slice(0, 0, num_contexts).max().item<int32_t>();
```
Comment (brnguyen2) [inline cpp/tensorrt_llm/thop/attentionOp.cpp]: This is character-for-character the same expression as line 901, which sets `enqueue_params.cross_kv_length` for the context run (same tensor, same `slice(0, 0, num_contexts)`). Keeping two copies in sync by convention is what produced the bug this PR fixes; consider hoisting it to a single `const` above the `getWorkspaceSize` call and passing it into `run()` (or at least a cross-reference comment on both).

Minor gating difference worth a note in the comment: line 901 additionally requires `cross_kv.has_value()`, so an `isCrossAttention()` op called without a `cross_kv` tensor gets a nonzero size here and a zero at enqueue. That direction is safe (over-allocates), but a reader checking the two for equality will trip on it.

Comment (pranav-nvidia) [reply cpp/tensorrt_llm/thop/attentionOp.cpp]: Kept the two computations but documented them in both directions, and called out the `cross_kv.has_value()` asymmetry as intentional safe over-allocation. I did try hoisting it and passing it into `run()`. It works and is strictly better, but it adds a parameter to that signature and reflows both call sites, which felt like scope creep here. Happy to fold it into the refactor follow-up.

File: cpp/tensorrt_llm/common/attentionOp.cpp (hunk 1 above)
Comment (pranav-nvidia) [reply cpp/tensorrt_llm/common/attentionOp.cpp:805]: Agreed on the shared `computeContextWorkspaceSizes`. The `TLLM_CHECK` needs a new `EnqueueParams` field too. It currently carries `void* workspace` (`attentionOp.h:119`) with no allocated byte count to compare against. I will take both as a follow-up PR rather than widening this one.

File: cpp/tensorrt_llm/thop/attentionOp.cpp [hunk 3]
```diff
@@ -410,10 +411,11 @@ class Runner : public RunnerBase
     }
 
     int64_t getWorkspaceSize(AttentionOp const& op, int const num_tokens, int const max_attention_window_size,
-        int const num_gen_tokens, int const max_blocks_per_sequence, int const ctx_total_kv_len = 0) const override
+        int const num_gen_tokens, int const max_blocks_per_sequence, int const ctx_total_kv_len = 0,
+        int const maxCrossKvLength = 0) const override
     {
         size_t const context_workspace_size = op.getWorkspaceSizeForContext(
```
Comment (pengbowang-nv) [inline cpp/tensorrt_llm/thop/attentionOp.cpp:417]: This function tend to take all maximum values at once: `max_batch_size` for `max_num_requests`, `min(max_seq_len, max_num_tokens)` for `op.mMaxContextLength`, even without `maxCrossKvLength`, that would be a very large overhead for memory. And the path is shared by decoder only models so it may get very large `seqlen` and `batchsize`, is this expected?

Comment (pranav-nvidia) [reply cpp/tensorrt_llm/thop/attentionOp.cpp:417]: @pengbowang-nv That's a really good point. It does seem like the existing policy can over-allocate. Since the workspace is resized on demand and multi-turn KV lengths are provided at runtime, it should be possible to size the context portion from the actual batch dimensions while retaining conservative generation sizing.

This may be out of scope for this PR which mainly focuses on the cross-attention correctness fix. I can track the broader over-allocation in an nvbug and file a followup PR. Let me know if you'd prefer this optimization to be included in this PR!

**Review by pengbowang-nv — APPROVED** [review]
LGTM. Note that we need a followup PR to reduce unnecessary memory cost. Thanks!

File: cpp/tensorrt_llm/thop/attentionOp.cpp (hunk 3 above)
Comment (pranav-nvidia) [reply cpp/tensorrt_llm/thop/attentionOp.cpp:417]: Following offline discussion, I will follow up with a PR to address the latent over-allocation behavior.
